ACL2022/Models/are_models

```diff
@@ -0,0 +1,324 @@
+               loss='MSE',
+               max_seq_len = 128,
+               embedding_size = 200,
+               monitor_loss = 'val_loss',
+               metrics = [MeanSquaredError(name="MSE"),
+                          MeanAbsoluteError(name="MAE"),
+                          MeanSquaredLogarithmicError(name="MSLE"),
+                          ]  
+              ):
+        self.lr = lr
+        self.batch_size = batch_size
+        self.rnn_layer_size = rnn_layer_size
+        self.attention_hidden_layers = 2
+        self.max_seq_len =  max_seq_len
+        self.show_summary = show_summary
+        self.patience=patience
+        self.epochs = epochs
+        self.loss = loss
+        self.embedding_size = embedding_size
+        self.monitor_loss = monitor_loss
+        self.metrics = metrics
+        self.earlystop = tf.keras.callbacks.EarlyStopping(monitor=self.monitor_loss,
+                                                          patience=self.patience,
+                                                          verbose=1,
+                                                          restore_best_weights=True,
+                                                          mode="min"
+                                                         )
+        self.unk_token = "[unk]"
+        self.pad_token = "[pad]"
+
+  #dummy loss for att weights (ignores them)
+  def dummy_loss(self, y_true, y_pred):
+      return 0.0
+
+  def build(self, vocab_size):
+      inputs= Input(shape=(self.max_seq_len,), name="inputs")
+      x = Embedding(input_dim=vocab_size, output_dim=self.embedding_size, 
+                      input_length=self.max_seq_len, mask_zero=True, trainable=True)(inputs)
+      rnn = Bidirectional(LSTM(self.rnn_layer_size, return_sequences=True))(x)
+      for i in range(self.attention_hidden_layers): #-1):
+        if i == 0:
+          x = TimeDistributed(Dense(self.rnn_layer_size, activation = None))(rnn)
+        elif i == self.attention_hidden_layers-1: #last layer
+          x = TimeDistributed(Dense(1, activation = None))(x)
+        else:
+          x = TimeDistributed(Dense(self.rnn_layer_size, activation = None))(x) #hidden layer
+      att_weights = tensorflow.nn.softmax(x, axis=1)
+      weighted_represantation = tensorflow.math.multiply(att_weights, rnn, name=None)
+      weighted_represantation = tensorflow.reduce_sum(weighted_represantation, axis = 1)
+      pred = Dense(1, activation = None, name = 'classification')(weighted_represantation)
+      self.model = Model(inputs=inputs, outputs=[pred, tensorflow.squeeze(att_weights)])
+      self.model.compile(loss=[self.loss, self.dummy_loss] ,
+                    optimizer=tensorflow.keras.optimizers.Adam(learning_rate=self.lr),
+                    metrics=self.metrics)
+      if self.show_summary:
+          self.model.summary()
+
+  def create_vocab(self, tokenized_texts):
+    self.vocab = list(set([w for txt in tokenized_texts for w in txt]))
```

Comment:
This is not a memory efficient way of doing this, you are creating a very large list before creating a set. Use a set comprehension instead.